src/Compilers/Core/Portable/Options/SyntaxTreeOptionsProvider.cs

```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.Options
+{
+    /// <summary>
+    /// A class that provides <see cref="OptionSet" /> instances for a given file path. Implementations of this should be
+    /// immutable; that is two implementations that are Equal according to <see cref="Object.Equals(object)"/> should return
+    /// the same options for the same path.
+    /// </summary>
+    public abstract class SyntaxTreeOptionsProvider
+    {
+        /// <summary>
+        /// Provides an <see cref="OptionSet"/> for options that apply to the given tree.
+        /// </summary>
+        /// <param name="path">The path of the syntax tree to provide options for.</param>
+        /// <returns>An <see cref="OptionSet"/>. Will never return null.</returns>
+        // TODO: determine if this path is post-pathmap'ed or if this function still needs to do path mapping itself
```

Comment:
I'll actually need to reword this since we'll also need options for additional files. I'd try to pretend I thought of that _before_ sending this PR out, but I can't do that with a straight face.